retail/retail-java-applications/terraform/test/pubsub/README.md

```diff
@@ -0,0 +1,114 @@
+# Automated Test For Terraform Modules
+
+Use Go and Terratest to test the following modules.
+- Pub/Sub
+
+## Install Terraform and Go
+
+If you are using the [Google Cloud Shell](https://cloud.google.com/shell/docs/how-cloud-shell-works), both Terraform and Go are installed.
+
+Follow the instructions to install [Terraform cli](https://learn.hashicorp.com/tutorials/terraform/install-cli?in=terraform/gcp-get-started) and [Go](https://golang.org/doc/install).
+
+This repo has been tested on Terraform version `0.14.5`, Go version `1.16` and Google provider version `3.48.0`
+
+## Terraform Folder Structure
+
+
+The Terraform folder structure can be found below. 
+
+Note that the `terraform` folder is inside `dataflow-sample-applications/retail/retail-java-applications` folder.
+
+```
+terraform
+├── examples
+│   └── pubsub
+│       ├── main.tf
+│       ├── terraform.tfvars
+│       └── variables.tf
+├── main.tf
+├── modules
+│   └── pubsub
+│       ├── main.tf
+│       └── variables.tf
+├── README.MD
+├── terraform.tfvars
+├── test
+│   └── pubsub
+│       ├── go.mod
+│       ├── go.sum
+│       ├── pubsub_test.go
+│       └── README.md
+└── variables.tf
+```
+
+The `examples` folder contains Terraform scripts that serve as living documentation about how to use the module. They are also used by automated tests for execution.
+
+The `modules` folder contains the scripts that form the basis of the module.
+
+The `test` folder contains the test code.
+
+The `terraform` folder contains Terraform scripts which deploy required resources for `retail java application` using corresponding Terraform modules.
+
+## Resources Being Tested
+
+### Pub/Sub
+The test will clean up conflicting resources if there are any. 
+
+After that, four topics and three subscriptions will be created in Pub/Sub with names described below. 
+
+**Topics**
+- Clickstream-inbound
+- Transactions-inbound
+- Inventory-inbound
+- Inventory-outbound
+
+**Subscriptions**
+- Clickstream-inbound-sub
+- Transactions-inbound-sub
+- Inventory-inbound-sub
+
+Once these resources are created successfully, the test will publish messages using the topics and receive messages using the subscriptions. 
+
+Messages sent and received must match in order to pass the test. 
+
+Finally, all resources will be destroyed in order to keep a clean test environment.
+
+
+## Export Your Project ID
+
+Please replace `[your project ID]` with your own project ID in the following command and execute it:
+
+`export GOOGLE_PROJECT=[your project ID]`
+
+The test will pick up the project ID from environment variables automatically.
+
+## Execute Automated Test
+
+
+1. Make sure you are in the test folder: `dataflow-sample-applications/retail/retail-java-applications/terraform/test`
+2. Execute the test using `go test -v`
```

Comment:
If I'm in this folder and run `go test -v`, I get:
```
go: cannot find main module, but found .git/config in /home/tth/projects/dataflow-sample-applications
        to create a module there, run:
        cd ../../../.. && go mod init
```

If I go into the `pubsub` folder and run `go test -v`, I get:
```
retry.go:99: Returning due to fatal error: FatalError{Underlying: error while running command: chdir ../examples/pubsub: no such file or directory; }
    destroy.go:11: 
                Error Trace:    destroy.go:11
                                                        panic.go:613
                                                        testing.go:742
                                                        testing.go:810
                                                        init.go:13
                                                        pubsub_test.go:45
                Error:          Received unexpected error:
                                FatalError{Underlying: error while running command: chdir ../examples/pubsub: no such file or directory; }
                Test:           TestPubsub
--- FAIL: TestPubsub (0.80s)
```

Reply:
My Apology, it was caused by change in relative path. The fix is done and tested. Thank you.